File: todo/c/toradora.c
/* toradora.c - A terminal based todo list */
/* Using a sqlite database at ~/.toradora.db we will keep track of all the
 * entries */

/** FEATURES **/
/*
 * 1. Able to add tasks. (toradora -a "The task I want to add")
 * 2. Able to complete those tasks. (toradora -c 1 )
 * 3. Able to edit those tasks. (toradora -e 1 "The previous task edited")
 * 4. Able to manifest those tasks. (toradora)
 * ===============================
 * >>> Daily <<<
 * [!] 1. "daily habit"
 *
 * >>> Today's Task <<<
 * [!] 2. "high priority task"
 * [~] 3. "normal task"
 *
 * >>> Backlog <<<
 * [ ] 4. "old task"
 * ===============================
 * 5. Able to manifest all the tasks in database. (toradora -z)
 * 6. Able to add a daily task. (toradora -d "daily habit")
 * 7. Able to add a task with priority. (toradora -a "task" -p <1|2|3>)
 *    Priority: 1 = high [!], 2 = medium [~], 3 = low [ ]
 *
 */

#include <sqlite3.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define VERSION "0.2.0"

#define MAX_PATH_LENGTH 4096

/* Priority levels */
#define PRIORITY_HIGH   1
#define PRIORITY_MEDIUM 2
#define PRIORITY_LOW    3

static const char *
priority_label(int priority)
{
    switch (priority) {
    case PRIORITY_HIGH:   return "[!]";
    case PRIORITY_MEDIUM: return "[~]";
    case PRIORITY_LOW:    return "[ ]";
    default:              return "[~]";
    }
}

static void
manifest_usage()
{
    printf("toradora - a terminal based todo list\n\n");
    printf("Usage:\n");
    printf("  toradora                          Show today's tasks and backlog\n");
    printf("  toradora -z                       Show all tasks ever recorded\n");
    printf("  toradora -a \"task\"                Add a new task\n");
    printf("  toradora -a \"task\" -p <1|2|3>     Add a task with priority (1=high, 2=medium, 3=low)\n");
    printf("  toradora -d \"task\"                Add a daily task (always shown)\n");
    printf("  toradora -c <id>                  Mark task <id> as complete\n");
    printf("  toradora -e <id> \"new task\"       Edit task <id>\n\n");
    printf("Priority levels:\n");
    printf("  1 = [!] High\n");
    printf("  2 = [~] Medium (default)\n");
    printf("  3 = [ ] Low\n\n");
    printf("Examples:\n");
    printf("  toradora -a \"write usage function\"\n");
    printf("  toradora -a \"urgent fix\" -p 1\n");
    printf("  toradora -d \"morning run\"\n");
    printf("  toradora -c 3\n");
    printf("  toradora -e 3 \"write usage function (done)\"\n");
}

/* Callback for displaying tasks with priority label */
static int
callback_manifest(void *data, int argc, char **argv, char **az_col_name)
{
    (void)data;
    (void)az_col_name;
    /* argv[0] = id, argv[1] = task, argv[2] = priority */
    int priority = PRIORITY_MEDIUM;
    if (argc >= 3 && argv[2] != NULL) {
        priority = atoi(argv[2]);
    }
    printf("  %s %s. %s\n", priority_label(priority), argv[0], argv[1]);
    return 0;
}

static void
manifest_active_tasks(sqlite3 *DB)
{
    int query_status;
    char *message_error_1;
    char *message_error_2;
    char *message_error_3;

    printf("----------------------------------------------\n");

    /* Daily tasks */
    printf(">>> Daily <<<\n");
    char *manifest_query_daily = "SELECT id, task, priority FROM tasks "
        "WHERE status = 0 AND daily = 1 ORDER BY priority ASC;";
    query_status = sqlite3_exec(DB, manifest_query_daily, callback_manifest, NULL, &message_error_1);
    if (query_status != SQLITE_OK) {
        printf("[Error]: Unexpected error during execution of query\n");
        printf("%s\n", message_error_1);
        printf("[Help]: Check database sanity\n");
        return;
    }

    /* Today's tasks (non-daily, created today) */
    printf("\n>>> Today's Tasks <<<\n");
    char *manifest_query_today = "SELECT id, task, priority FROM tasks "
        "WHERE status = 0 AND daily = 0 AND created_at = CURRENT_DATE "
        "ORDER BY priority ASC;";
    query_status = sqlite3_exec(DB, manifest_query_today, callback_manifest, NULL, &message_error_2);
    if (query_status != SQLITE_OK) {
        printf("[Error]: Unexpected error during execution of query\n");
        printf("%s\n", message_error_2);
        printf("[Help]: Check database sanity\n");
        return;
    }

    /* Backlog (non-daily, created before today) */
    printf("\n>>> Backlog <<<\n");
    char *manifest_query_backlog = "SELECT id, task, priority FROM tasks "
        "WHERE status = 0 AND daily = 0 AND created_at != CURRENT_DATE "
        "ORDER BY priority ASC;";
    query_status = sqlite3_exec(DB, manifest_query_backlog, callback_manifest, NULL, &message_error_3);
    if (query_status != SQLITE_OK) {
        printf("[Error]: Unexpected error during execution of query\n");
        printf("%s\n", message_error_3);
        printf("[Help]: Check database sanity\n");
        return;
    }

    printf("----------------------------------------------\n");
}

static void
manifest_all_tasks(sqlite3 *DB)
{
    int query_status;
    char *message_error;
    printf("----------------------------------------------\n");
    printf(">>> All tasks ever recorded <<<\n");
    char *manifest_query = "SELECT id, task, priority FROM tasks ORDER BY priority ASC;";
    query_status = sqlite3_exec(DB, manifest_query, callback_manifest, NULL, &message_error);
    if (query_status != SQLITE_OK) {
        printf("[Error]: Unexpected error during execution of query\n");
        printf("%s\n", message_error);
        printf("[Help]: Check database sanity\n");
        return;
    }
    printf("----------------------------------------------\n");
}

static void
toradora_add_task(sqlite3 *DB, const char *todo, int priority)
{
    sqlite3_stmt *stmt = NULL;
    int query_status;

    const char *add_query = "INSERT INTO tasks (task, priority) VALUES (?, ?);";

    query_status = sqlite3_prepare_v2(DB, add_query, -1, &stmt, NULL);
    if (query_status != SQLITE_OK) {
        printf("[Error]: Unexpected error during execution of query\n");
        printf("[Help]: Check database sanity\n");
        goto error;
    }

    sqlite3_bind_text(stmt, 1, todo, -1, SQLITE_STATIC);
    sqlite3_bind_int(stmt, 2, priority);
    query_status = sqlite3_step(stmt);
    if (query_status != SQLITE_DONE) {
        printf("[Error]: Unexpected error during execution of query\n");
        printf("[Help]: Check database sanity\n");
        goto error;
    }
    printf("[Info]: Task added with priority %s\n", priority_label(priority));
    error:
    sqlite3_finalize(stmt);
}

static void
toradora_add_daily_task(sqlite3 *DB, const char *todo, int priority)
{
    sqlite3_stmt *stmt = NULL;
    int query_status;

    const char *add_query = "INSERT INTO tasks (task, priority, daily) VALUES (?, ?, 1);";

    query_status = sqlite3_prepare_v2(DB, add_query, -1, &stmt, NULL);
    if (query_status != SQLITE_OK) {
        printf("[Error]: Unexpected error during execution of query\n");
        printf("[Help]: Check database sanity\n");
        goto error;
    }

    sqlite3_bind_text(stmt, 1, todo, -1, SQLITE_STATIC);
    sqlite3_bind_int(stmt, 2, priority);
    query_status = sqlite3_step(stmt);
    if (query_status != SQLITE_DONE) {
        printf("[Error]: Unexpected error during execution of query\n");
        printf("[Help]: Check database sanity\n");
        goto error;
    }
    printf("[Info]: Daily task added with priority %s\n", priority_label(priority));
    error:
    sqlite3_finalize(stmt);
}

static void
toradora_complete_task(sqlite3 *DB, const char *id)
{
    sqlite3_stmt *stmt = NULL;
    int query_status;
    int curr;
    char *end_ptr;
    int row_affected;
    const char *complete_query = "UPDATE tasks SET status = 1 WHERE id = ?;";

    query_status = sqlite3_prepare_v2(DB, complete_query, -1, &stmt, NULL);
    if (query_status != SQLITE_OK) {
        printf("[Error]: Unexpected error during execution of query\n");
        printf("[Help]: Check database sanity\n");
        goto error;
    }
    curr = (int)strtol(id, &end_ptr, 10);
    if (*end_ptr != '\0') {
        printf("[Error]: Invalid argument\n");
        manifest_usage();
        goto error;
    }
    sqlite3_bind_int(stmt, 1, curr);
    query_status = sqlite3_step(stmt);
    if (query_status != SQLITE_DONE) {
        printf("[Error]: Unexpected error during execution of query\n");
        printf("[Help]: Check database sanity\n");
        goto error;
    }
    row_affected = sqlite3_changes(DB);
    if (row_affected == 0) {
        printf("[Info]: No rows were affected\n");
        printf("[Help]: Did you enter a valid task id?\n");
    } else {
        printf("[Info]: Task %d marked as complete\n", curr);
    }
    error:
    sqlite3_finalize(stmt);
}

static void
toradora_edit_task(sqlite3 *DB, const char *id, const char *new_task)
{
    sqlite3_stmt *stmt = NULL;
    int query_status;
    int curr;
    char *end_ptr;
    int row_affected;
    const char *edit_query = "UPDATE tasks SET task = ? WHERE id = ?;";

    query_status = sqlite3_prepare_v2(DB, edit_query, -1, &stmt, NULL);
    if (query_status != SQLITE_OK) {
        printf("[Error]: Unexpected error during execution of query\n");
        printf("[Help]: Check database sanity\n");
        goto error;
    }
    curr = (int)strtol(id, &end_ptr, 10);
    if (*end_ptr != '\0') {
        printf("[Error]: Invalid argument\n");
        manifest_usage();
        goto error;
    }
    sqlite3_bind_text(stmt, 1, new_task, -1, SQLITE_STATIC);
    sqlite3_bind_int(stmt, 2, curr);
    query_status = sqlite3_step(stmt);
    if (query_status != SQLITE_DONE) {
        printf("[Error]: Unexpected error during execution of query\n");
        printf("[Help]: Check database sanity\n");
        goto error;
    }
    row_affected = sqlite3_changes(DB);
    if (row_affected == 0) {
        printf("[Info]: No rows were affected\n");
        printf("[Help]: Did you enter a valid task id?\n");
    } else {
        printf("[Info]: Task %d updated\n", curr);
    }
    error:
    sqlite3_finalize(stmt);
}

/* Ensure new columns exist on older databases (migration) */
static void
toradora_migrate_db(sqlite3 *DB)
{
    char *err = NULL;
    /* Ignore errors - columns may already exist */
    sqlite3_exec(DB, "ALTER TABLE tasks ADD COLUMN priority INTEGER NOT NULL DEFAULT 2;", NULL, 0, &err);
    if (err) sqlite3_free(err);
    err = NULL;
    sqlite3_exec(DB, "ALTER TABLE tasks ADD COLUMN daily INTEGER NOT NULL DEFAULT 0;", NULL, 0, &err);
    if (err) sqlite3_free(err);
}

int
main(int argc, char *argv[])
{
    char *home_path = getenv("HOME");
    char *file_path = "/.toradora.db";
    char database_path[MAX_PATH_LENGTH];
    sqlite3 *DB;
    int db_exit = 0;
    int tab_exit = 0;
    char *message_error;

    if (home_path == NULL) {
        printf("[Error]: HOME environment variable not set.\n");
        return -1;
    }
    snprintf(database_path, sizeof(database_path), "%s%s", home_path, file_path);

    db_exit = sqlite3_open(database_path, &DB);
    if (db_exit) {
        printf("[Error]: Database corrupted\n");
        printf("[Help]: Delete database file and rerun the application\n");
        return -1;
    }

    char *create_table = "CREATE TABLE IF NOT EXISTS tasks ("
        "id INTEGER PRIMARY KEY NOT NULL,"
        "task TEXT NOT NULL,"
        "status INTEGER NOT NULL DEFAULT 0,"
        "priority INTEGER NOT NULL DEFAULT 2,"
        "daily INTEGER NOT NULL DEFAULT 0,"
        "created_at TEXT NOT NULL DEFAULT CURRENT_DATE"
        ");";

    tab_exit = sqlite3_exec(DB, create_table, NULL, 0, &message_error);
    if (tab_exit != SQLITE_OK) {
        printf("[Error]: Table creation failed\n");
        printf("%s\n", message_error);
        return -1;
    }

    /* Migrate existing databases to add new columns if missing */
    toradora_migrate_db(DB);

    /*
     * Argument parsing:
     *   toradora                          -> show active tasks
     *   toradora -z                       -> show all tasks
     *   toradora -h                       -> show usage
     *   toradora -a "task"                -> add task (medium priority)
     *   toradora -a "task" -p <1|2|3>    -> add task with priority
     *   toradora -d "task"               -> add daily task (medium priority)
     *   toradora -d "task" -p <1|2|3>   -> add daily task with priority
     *   toradora -c <id>                 -> complete task
     *   toradora -e <id> "new task"      -> edit task
     */
    if (argc == 1) {
        manifest_active_tasks(DB);
    } else if (argc == 2) {
        if (0 == strcmp(argv[1], "-z")) {
            manifest_all_tasks(DB);
        } else if (0 == strcmp(argv[1], "-h")) {
            manifest_usage();
        } else {
            manifest_usage();
            sqlite3_close(DB);
            return 1;
        }
    } else if (argc == 3) {
        if (0 == strcmp(argv[1], "-a")) {
            toradora_add_task(DB, argv[2], PRIORITY_MEDIUM);
        } else if (0 == strcmp(argv[1], "-d")) {
            toradora_add_daily_task(DB, argv[2], PRIORITY_MEDIUM);
        } else if (0 == strcmp(argv[1], "-c")) {
            toradora_complete_task(DB, argv[2]);
        } else {
            manifest_usage();
            sqlite3_close(DB);
            return 1;
        }
    } else if (argc == 4) {
        if (0 == strcmp(argv[1], "-e")) {
            toradora_edit_task(DB, argv[2], argv[3]);
        } else {
            manifest_usage();
            sqlite3_close(DB);
            return 1;
        }
    } else if (argc == 5) {
        /* toradora -a "task" -p <1|2|3> */
        /* toradora -d "task" -p <1|2|3> */
        if ((0 == strcmp(argv[1], "-a") || 0 == strcmp(argv[1], "-d"))
            && 0 == strcmp(argv[3], "-p")) {
            char *end_ptr;
            int priority = (int)strtol(argv[4], &end_ptr, 10);
            if (*end_ptr != '\0' || priority < 1 || priority > 3) {
                printf("[Error]: Priority must be 1, 2, or 3\n");
                manifest_usage();
                sqlite3_close(DB);
                return 1;
            }
            if (0 == strcmp(argv[1], "-a")) {
                toradora_add_task(DB, argv[2], priority);
            } else {
                toradora_add_daily_task(DB, argv[2], priority);
            }
        } else {
            manifest_usage();
            sqlite3_close(DB);
            return 1;
        }
    } else {
        manifest_usage();
        sqlite3_close(DB);
        return 1;
    }

    sqlite3_close(DB);
    return 0;
}

File: todo/python/toradora.py
#!/usr/bin/env python3
"""Small SQLite-backed command-line todo list.

The store in this module is also used by the Flask implementation.  Keeping
database operations here means both Python tracks exercise the same behavior.
"""

from __future__ import annotations

import argparse
import sqlite3
from dataclasses import dataclass
from datetime import date
from pathlib import Path
from typing import Iterable


PRIORITY_LABELS = {1: "[!]", 2: "[~]", 3: "[ ]"}
DEFAULT_DB_PATH = Path.home() / ".toradora.db"


@dataclass(frozen=True)
class Task:
    id: int
    task: str
    status: int
    priority: int
    long_term: bool
    created_at: str

    @classmethod
    def from_row(cls, row: sqlite3.Row) -> "Task":
        return cls(
            id=row["id"],
            task=row["task"],
            status=row["status"],
            priority=row["priority"],
            long_term=bool(row["daily"]),
            created_at=row["created_at"],
        )


class TodoStore:
    """SQLite persistence for Toradora tasks."""

    def __init__(self, database_path: str | Path = DEFAULT_DB_PATH):
        self.database_path = Path(database_path).expanduser()
        self.database_path.parent.mkdir(parents=True, exist_ok=True)
        self._initialise()

    def _connect(self) -> sqlite3.Connection:
        connection = sqlite3.connect(self.database_path)
        connection.row_factory = sqlite3.Row
        return connection

    def _initialise(self) -> None:
        with self._connect() as connection:
            connection.execute(
                """CREATE TABLE IF NOT EXISTS tasks (
                    id INTEGER PRIMARY KEY,
                    task TEXT NOT NULL,
                    status INTEGER NOT NULL DEFAULT 0,
                    priority INTEGER NOT NULL DEFAULT 2,
                    daily INTEGER NOT NULL DEFAULT 0,
                    created_at TEXT NOT NULL DEFAULT CURRENT_DATE
                )"""
            )
            # Keep databases made by the original C app usable.
            columns = {
                row["name"]
                for row in connection.execute("PRAGMA table_info(tasks)")
            }
            if "priority" not in columns:
                connection.execute(
                    "ALTER TABLE tasks ADD COLUMN priority INTEGER NOT NULL DEFAULT 2"
                )
            if "daily" not in columns:
                connection.execute(
                    "ALTER TABLE tasks ADD COLUMN daily INTEGER NOT NULL DEFAULT 0"
                )
            if "created_at" not in columns:
                connection.execute(
                    "ALTER TABLE tasks ADD COLUMN created_at TEXT NOT NULL DEFAULT CURRENT_DATE"
                )

    def add(self, task: str, priority: int = 2, long_term: bool = False) -> Task:
        task = task.strip()
        if not task:
            raise ValueError("Task text cannot be empty")
        if priority not in PRIORITY_LABELS:
            raise ValueError("Priority must be 1, 2, or 3")
        with self._connect() as connection:
            cursor = connection.execute(
                "INSERT INTO tasks (task, priority, daily) VALUES (?, ?, ?)",
                (task, priority, int(long_term)),
            )
            row = connection.execute(
                "SELECT * FROM tasks WHERE id = ?", (cursor.lastrowid,)
            ).fetchone()
        return Task.from_row(row)

    def active(self) -> list[Task]:
        with self._connect() as connection:
            rows = connection.execute(
                """SELECT * FROM tasks
                   WHERE status = 0
                   ORDER BY daily DESC, priority ASC, id ASC"""
            ).fetchall()
        return [Task.from_row(row) for row in rows]

    def all(self) -> list[Task]:
        with self._connect() as connection:
            rows = connection.execute(
                "SELECT * FROM tasks ORDER BY priority ASC, id ASC"
            ).fetchall()
        return [Task.from_row(row) for row in rows]

    def complete(self, task_id: int) -> bool:
        with self._connect() as connection:
            cursor = connection.execute(
                "UPDATE tasks SET status = 1 WHERE id = ?", (task_id,)
            )
        return cursor.rowcount == 1

    def edit(self, task_id: int, task: str) -> bool:
        task = task.strip()
        if not task:
            raise ValueError("Task text cannot be empty")
        with self._connect() as connection:
            cursor = connection.execute(
                "UPDATE tasks SET task = ? WHERE id = ?", (task, task_id)
            )
        return cursor.rowcount == 1

    def delete(self, task_id: int) -> bool:
        with self._connect() as connection:
            cursor = connection.execute("DELETE FROM tasks WHERE id = ?", (task_id,))
        return cursor.rowcount == 1


def priority_label(priority: int) -> str:
    return PRIORITY_LABELS.get(priority, "[~]")


def grouped_active_tasks(tasks: Iterable[Task]) -> dict[str, list[Task]]:
    today = date.today().isoformat()
    groups = {"Long-term": [], "Today's Tasks": [], "Backlog": []}
    for task in tasks:
        if task.long_term:
            groups["Long-term"].append(task)
        elif task.created_at == today:
            groups["Today's Tasks"].append(task)
        else:
            groups["Backlog"].append(task)
    return groups


def print_task(task: Task) -> None:
    marker = "[x]" if task.status else priority_label(task.priority)
    print(f"  {marker} {task.id}. {task.task}")


def print_active(tasks: Iterable[Task]) -> None:
    print("----------------------------------------------")
    for name, grouped in grouped_active_tasks(tasks).items():
        print(f">>> {name} <<<")
        for task in grouped:
            print_task(task)
        print()
    print("----------------------------------------------")


def print_all(tasks: Iterable[Task]) -> None:
    print("----------------------------------------------")
    print(">>> All tasks ever recorded <<<")
    for task in tasks:
        print_task(task)
    print("----------------------------------------------")


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="A small terminal todo list")
    parser.add_argument(
        "--db", type=Path, default=DEFAULT_DB_PATH, help="SQLite path (default: ~/.toradora.db)"
    )
    actions = parser.add_mutually_exclusive_group()
    actions.add_argument("-a", "--add", metavar="TASK", help="add a task")
    actions.add_argument(
        "-d", "--daily", "--long-term", dest="long_term", metavar="TASK",
        help="add a long-term task (always shown)",
    )
    actions.add_argument("-c", "--complete", type=int, metavar="ID", help="complete a task")
    actions.add_argument(
        "-e", "--edit", nargs=2, metavar=("ID", "TASK"), help="edit a task"
    )
    actions.add_argument("--delete", type=int, metavar="ID", help="delete a task")
    actions.add_argument("-z", "--all", action="store_true", help="show all tasks")
    parser.add_argument(
        "-p", "--priority", type=int, choices=(1, 2, 3), default=2,
        help="1=high, 2=medium, 3=low (default: 2)",
    )
    return parser


def main(argv: list[str] | None = None) -> int:
    parser = build_parser()
    args = parser.parse_args(argv)
    store = TodoStore(args.db)

    try:
        if args.add is not None:
            task = store.add(args.add, args.priority)
            print(f"Task {task.id} added with priority {priority_label(task.priority)}")
        elif args.long_term is not None:
            task = store.add(args.long_term, args.priority, long_term=True)
            print(f"Long-term task {task.id} added with priority {priority_label(task.priority)}")
        elif args.complete is not None:
            print(
                f"Task {args.complete} marked as complete"
                if store.complete(args.complete)
                else "No task found with that id"
            )
        elif args.edit is not None:
            task_id, text = args.edit
            if store.edit(int(task_id), text):
                print(f"Task {task_id} updated")
            else:
                print("No task found with that id")
        elif args.delete is not None:
            print(
                f"Task {args.delete} deleted"
                if store.delete(args.delete)
                else "No task found with that id"
            )
        elif args.all:
            print_all(store.all())
        else:
            print_active(store.active())
    except (ValueError, sqlite3.Error) as error:
        parser.error(str(error))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
